ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/SourceLocationManager.java

```diff
@@ -441,11 +461,43 @@ private static SourceExtensions processExtensions() {
 	protected BundleManifestSourceLocationManager initializeBundleManifestLocations() {
 		BundleManifestSourceLocationManager manager = new BundleManifestSourceLocationManager();
 		manager.setPlugins(PDECore.getDefault().getModelManager().getExternalModels());
+		try {
+			ITargetDefinition definition = TargetPlatformService.getDefault().getWorkspaceTargetDefinition();
+			if (definition != null && definition.isResolved()) {
+				TargetBundle[] bundles = definition.getAllBundles();
+				if (bundles != null) {
+					manager.setTargetBundles(bundles);
+				}
+			}
+		} catch (CoreException e) {
+			// can't use this then...
+		}
 		return manager;
 	}
 
 	private static final class SourceExtensions {
 		final Collection<SourceLocation> locations = new LinkedHashSet<>();
-		final List<IPluginSourcePathLocator> locators = new ArrayList<>();
+		final List<OrderedPluginSourcePathLocator> locators = new ArrayList<>();
+	}
+
+	private static final class OrderedPluginSourcePathLocator implements Comparable<OrderedPluginSourcePathLocator> {
+
+		private IPluginSourcePathLocator locator;
+		private LocatorComplexity complexity;
+
+		OrderedPluginSourcePathLocator(IPluginSourcePathLocator locator, LocatorComplexity complexity) {
+			this.locator = locator;
+			this.complexity = complexity;
+		}
+
+		@Override
+		public int compareTo(OrderedPluginSourcePathLocator o) {
+			return complexity.compareTo(o.complexity);
+		}
+
+	}
+
+	private enum LocatorComplexity {
+		low, medium, high, unkown;
```

Comment:
Is the concept of complexity still relevant with your latest change?
If not I suggest to remove it again.

Reply:
Yes for example m2e will most probably have a mediom and/or high one implementation.

Reply:
Understand 👍🏽
I think it would then be good to split the commit into two. One that introduces the complexity and one that adds support for 'native target-sources'.
You probably also want to update the commit message of this that commit.